onnxscript/optimizer/_constant_folding.py

```diff
@@ -999,31 +1005,45 @@ def process_node(self, node: ir.Node) -> Replacement | None:
                     output = [output]
                 return Replacement(output, context.nodes)
 
-        if is_control_flow_op(node) or is_non_deterministic_op(node):
+        if _is_control_flow_op(node) or _is_non_deterministic_op(node):
             return None
 
-        if is_onnx_op(node, "Constant"):
+        if _is_onnx_op(node, "Constant"):
             _process_constant_node(node)
             return None
 
-        input_values = [_get_numpy_value(x) for x in node.inputs]
-        if any(x is None for x in input_values):
+        if any(x.is_graph_input() for x in node.inputs if x is not None):
+            # Do not fold any graph inputs to preserve graph signature
             return None
 
-        if any(self._state.is_initializer_input(x) for x in node.inputs):  # type: ignore[arg-type]
+        # Ensure all node inputs are constants
+        if any(x.const_value is None for x in node.inputs if x is not None):
+            if logger.isEnabledFor(logging.DEBUG):
+                logger.debug(
+                    "Skipping constant folding for node %s because it has non-constant inputs",
+                    node,
+                    [x.name for x in node.inputs if x is not None],
+                )
             return None
 
-        if any(input.nbytes > self._input_size_limit for input in input_values):  # type: ignore[union-attr]
+        input_tensors = [x.const_value if x is not None else None for x in node.inputs]
+
+        if (node.domain, node.op_type) not in self.always_fold_ops and any(
```

Comment:
Should always-fold-ups be restricted to values with a single-use? Eg, If we have a large initializer X that is used in two places, once Tranposed and once directly. Folding will double the memory-used for this X.